src/shared/parse-helpers.c

```diff
@@ -196,3 +202,129 @@ int parse_socket_bind_item(
+
+        if (r == 0)
+                return -EINVAL;
+
+        /* Enforce that at most 3 colon-separated words are present */
+        if (!isempty(v))
+                return -EINVAL;
+
+        for (const char *p = options;;) {
+                OpenFileFlags flag;
+                _cleanup_free_ char *word = NULL;
+
+                r = extract_first_word(&p, &word, ",", 0);
+                if (r < 0)
+                        return r;
+
+                if (r == 0)
+                        break;
+
+                flag = open_file_flags_from_string(word);
+                if (flag < 0)
+                        return flag;
+
+                if ((flag & ~_OPENFILE_MASK_PUBLIC) != 0)
+                        return -EINVAL;
+
+                of->flags |= flag;
+        }
+
+        r = open_file_check(of);
+        if (r < 0)
+                return r;
+
+        if (isempty(of->fdname)) {
+                free(of->fdname);
+                r = path_extract_filename(of->path, &of->fdname);
+                if (r < 0)
+                        return r;
+        } else if (!fdname_is_valid(of->fdname))
+                return -EINVAL;
+
+        *ret = TAKE_PTR(of);
+
+        return 0;
+}
+
+int open_file_check(const OpenFile *of) {
+        if (!path_is_valid(of->path) || !path_is_absolute(of->path))
+                return -EINVAL;
+
+        if ((!!FLAGS_SET(of->flags, OPENFILE_RDONLY) + !!FLAGS_SET(of->flags, OPENFILE_APPEND) +
+             !!FLAGS_SET(of->flags, OPENFILE_TRUNC)) > 1)
+                return -EINVAL;
+
+        return 0;
+}
+
+int open_file_to_string(const OpenFile *of, char **ret) {
+        _cleanup_free_ char *options = NULL;
+        _cleanup_free_ char *fname = NULL;
+        _cleanup_free_ char *s = NULL;
+        bool has_fdname = false;
+        int r;
+
+        assert(of);
+        assert(ret);
+
+        s = shell_escape(of->path, ":");
+        if (!s)
+                return -ENOMEM;
+
+        r = path_extract_filename(of->path, &fname);
+        if (r < 0)
+                return r;
+
+        has_fdname = !streq(fname, of->fdname);
+        if (has_fdname)
+                if (!strextend(&s, ":", of->fdname))
+                        return -ENOMEM;
+
+        for (OpenFileFlags flag = OPENFILE_RDONLY; flag < _OPENFILE_MAX; flag <<= 1)
+                if (FLAGS_SET(of->flags, flag) && !strextend_with_separator(&options, ",", open_file_flags_to_string(flag)))
+                        return -ENOMEM;
+
+        if (options)
+                if (!(has_fdname ? strextend(&s, ":", options) : strextend(&s, "::", options)))
+                        return -ENOMEM;
+
+        *ret = TAKE_PTR(s);
+
+        return 0;
+}
+
+OpenFile* open_file_free(OpenFile *of) {
+        if (!of)
+                return NULL;
+
+        free(of->path);
+        free(of->fdname);
+        return mfree(of);
+}
+
+static const char * const open_file_flags_table[_OPENFILE_MAX] = {
+        [OPENFILE_RDONLY] = "ro",
```

Comment:
please name the internal and the external names the same (modulo casing). i.e. if you call the internal thing "RDONLY" then the external one should be "rdonly", not "rd"

Reply:
I think in most cases we named similar options READ_ONLY/read-only btw, hence neither "ro" nor "rdonly", maybe stick to that